Source/Particles/Gather/FieldGather.H

```diff
@@ -0,0 +1,171 @@
+                    const amrex::Array4<const amrex::Real>& bz_arr,
+                    const long np_to_gather, 
+                    const std::array<amrex::Real,3>& dx,
+                    const std::array<Real, 3> xyzmin,
+                    const Dim3 lo,
+                    const amrex::Real stagger_shift)
+{
+    const Real dxi = 1.0/dx[0];
+    const Real dzi = 1.0/dx[2];
+#if (AMREX_SPACEDIM == 3)
+    const Real dyi = 1.0/dx[1];
+#endif
+
+    const Real xmin = xyzmin[0];
+    const Real ymin = xyzmin[1];
+    const Real zmin = xyzmin[2];
+
+    // Loop over particles and gather fields from {e,b}{x,y,z}_arr to {E,B}{xyz}p.
+    ParallelFor( np_to_gather,
+                 [=] AMREX_GPU_DEVICE (long ip) {
+                     // --- Compute shape factors
+                     // x direction
+                     // Get particle position after 1/2 push back in position
+                     const Real xmid = (xp[ip]-xmin)*dxi;
+                     // Compute shape factors for node-centered quantities
+                     Real AMREX_RESTRICT sx [depos_order + 1];
+                     // j: leftmost grid point (node-centered) that the particle touches
+                     const int j  = compute_shape_factor<depos_order>(sx, xmid);
+                     // Compute shape factors for cell-centered quantities
+                     Real AMREX_RESTRICT sx0[depos_order + 1 - lower_in_v];
+                     // j0: leftmost grid point (cell-centered) that the particle touches
+                     const int j0 = compute_shape_factor<depos_order - lower_in_v>(sx0, xmid-stagger_shift);
+                     
+#if (AMREX_SPACEDIM == 3)
+                     // y direction
+                     const Real ymid= (yp[ip]-ymin)*dyi;
+                     Real AMREX_RESTRICT sy [depos_order + 1];
+                     const int k  = compute_shape_factor<depos_order>(sy,  ymid);
+                     Real AMREX_RESTRICT sy0[depos_order + 1 - lower_in_v];
+                     const int k0 = compute_shape_factor<depos_order-lower_in_v>(sy0, ymid-stagger_shift);
+#endif
+                     // z direction
+                     const Real zmid= (zp[ip]-zmin)*dzi;
+                     Real AMREX_RESTRICT sz [depos_order + 1];
+                     const int l  = compute_shape_factor<depos_order>(sz,  zmid);
+                     Real AMREX_RESTRICT sz0[depos_order + 1 - lower_in_v];
+                     const int l0 = compute_shape_factor<depos_order - lower_in_v>(sz0, zmid-stagger_shift);
+
+                     // Deposit current into jx_arr, jy_arr and jz_arr
+                     Exp[ip] = 0;
+                     Eyp[ip] = 0;
+                     Ezp[ip] = 0;
+                     Bxp[ip] = 0;
+                     Byp[ip] = 0;
+                     Bzp[ip] = 0;
+#if (AMREX_SPACEDIM == 2)
+                     for (int iz=0; iz<=depos_order; iz++){
+                         for (int ix=0; ix<=depos_order; ix++){
```

Comment:
since those are fixed size, we could try adding some unroll pragmas in the inner (small) loops.

Reply:
Right, but aren't these pragmas compiler-dependent?
For GCC, do you suggest something like 
```
#pragma GCC unroll depos_order
```
?

Reply:
I was actually looking for ways to unroll these loops when I initially wrote this piece of code, so I am glad you bring it up :)

Reply:
Because `depos_order` is known to the compiler, I suspect it has enough information to decide what's best for performance.

Reply:
Single-GPU tests on summit with order 3 deposition for uniform plasma (I ran each test 3 times for statistics)

| no unroll | unroll inner | unroll all|
| -------- | --- | -- |
|0.799|0.7962|0.7896|
|0.7995|0.7959|0.7941|
|0.7962|0.7954|0.7908|

I'd say the game is not worth the candle, so let's not put NVCC's `#pragma unroll` in this piece of code.

Reply:
Yep, they are compiler specific and one could add a define cascade for supporting compilers and otherwise keep it empty.

Just an idea, how much did the kernel you benchmark contribute to the overall runtime? With regards that we are currently optimizing larger running kernels, the current improvement of 0.5% runtime could be revisited later on, if we like to.

Reply:
The numbers above are just for this field gather (I added a `cudaDeviceSynchronize()` after `ParallelFor()` for the sake of timing). So I'd say it is negligible and let's keep it this way.